### PR TITLE
streams: graph stage

### LINKED ISSUE #99: Replace ActorPublisher with GraphStage.
Hi,

In my Scala build, I'm currently getting a deprecation warning:

`object ActorPublisher` in package actor is deprecated (since 2.5.0): Use `akka.stream.stage.GraphStage` instead, it allows for all operations an Actor would and is more type-safe as well as guaranteed to be ReactiveStreams compliant.

We should therefore replace the `ActorPublisher`s with `GraphStage`.

### PR DESCRIPTION
**Pull request by ahjohannessen**
- add `StreamSource` and `AllStreamsSource` graph stages
   that replace deprecated `ActorPublisher` implementations
   `StreamPublisher` and `AllStreamsPublisher`.

Closes #99

Co-authored-by: bardurdam <bardurdam@gmail.com>

### CODE REVIEW

**Comment by ahjohannessen** [comment]
@t3hnar WDYT?

**Comment by t3hnar** [comment]
Looks nice, but I would need some time to review.
Ideally we need to rewrite `ConnectionActor` in therms of streams for better throughput and back-pressure  support

**Comment by ahjohannessen** [comment]
@t3hnar Yep, however this is a step in the right direction :) Ideally, we should eliminate usage of actors in the user facing API and strive for a more high-level principled approach using streams.

Long-term, we should put focus on separating the library into something a la:

 - `eventstore-client-core`: no akka dependency
 - `eventstore-client`: akka, akka-streams etc.

This separation would make it more feasible to provide alternative client API's, e.g. using fs2 / cats-effect / scalaz-effect / monix. However, this is for another talk and issue.

**Comment by ahjohannessen** [comment]
@t3hnar Have you had time to review? :)

File: src/main/scala/eventstore/AllStreamsSourceStage.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import akka.actor.ActorRef
+import akka.actor.Actor.Receive
+import akka.stream._
+import akka.stream.stage._
+import akka.stream.{ Attributes, SourceShape }
+import eventstore.ReadDirection.Forward
+import Position._
+import EventStream._
+
+class AllStreamsSourceStage(
```
Comment (t3hnar) [inline src/main/scala/eventstore/AllStreamsSourceStage.scala]: private ?

File: src/main/scala/eventstore/SourceStageLogic.scala [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+import akka.actor.ActorRef
+import akka.actor.Actor.Receive
+import akka.actor.Terminated
+import akka.actor.Status.Failure
+import akka.stream._
+import akka.stream.stage._
+import akka.stream.stage.GraphStageLogic.StageActor
+import akka.stream.SourceShape
+
+abstract class SourceStageLogic[T, O <: Ordered[O], P <: O](
```
Comment (t3hnar) [inline src/main/scala/eventstore/SourceStageLogic.scala]: private ?

File: src/main/scala/eventstore/StreamSourceStage.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import akka.actor.ActorRef
+import akka.actor.Actor.Receive
+import akka.actor.Status.Failure
+import akka.stream._
+import akka.stream.stage._
+import akka.stream.{ Attributes, SourceShape }
+import eventstore.ReadDirection.Forward
+import EventNumber._
+
+class StreamSourceStage(
```
Comment (t3hnar) [inline src/main/scala/eventstore/StreamSourceStage.scala]: private ?

**Pull request merged**